transformations/synonym_substitution/transformation.py

```diff
@@ -0,0 +1,68 @@
+import random
+import stanza
+
+stanza.download("en")
+import nltk
+
+nltk.download("wordnet")
+from nltk.corpus import wordnet
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+def synonym_substitution(text, stanza_pipeline, seed=42, prob=0.5):
+    random.seed(seed)
+    upos_wn_dict = {
+        "VERB": "v",
+        "NOUN": "n",
+        "ADV": "r",
+        "ADJ": "s",
+    }
+    result = []
+    doc = stanza_pipeline(text)
+    for sent in doc.sentences:
+        for word_dict in sent.words:
+            word = word_dict.text
+            wn_pos = upos_wn_dict.get(word_dict.upos)
```

Comment:
Maybe it's better to use the more detailed `xpos` here? I think the current implementation might result in some non-grammatical segments, such as the first test case (returned -> come).
```
{
  "type": "synonym_substitution",
  "test_cases": [
    {
      "class": "SynonymSubstitution",
      "inputs": {
        "sentence": "Andrew finally returned the French book to Chris that I bought last week"
      },
      "outputs": {
        "sentence": "Andrew finally come back the french book to Chris that I bought last workweek"
      }
```

Is it possible to add some sanity check to ensure that the original word and the synonym word have the same POS tag? I think that would make this generated text more natural. One way I can think of is to run some stanza pass on the transformed text with candidate synonym words and check if the POS tags for original/synonym words are the same. Then you could sample from those synonyms that matches the POS tags.

Maybe that would also make this transformation heavier :thinking:

Reply:
Good point. The bottomneck here is the synonyms from NLTK/worknet, which do not come with XPOS labels. I think it is doable to do a second round check, but as you mentioned that may make things more complicated, and more importantly, may introduce additional noises as no POS tagger is perfect. Currently, the synonyms will not be substituted if the UPOS doesn't match, which, in most of the cases, shouldn't affect human understanding of the sentences. How do you think about it?

Reply:
It looks fine to me. Maybe you could mention in the readme file that this transformation might introduce non-grammatical segments.

Reply:
Added. Thanks @Nickeilf !